src/github-graphql-base.js

```diff
@@ -0,0 +1,600 @@
+// sample usages:
+// 1. fetchAllIssues = await Client.getGraphQLClient()
+//      .repo(repoOwner, repoName)
+//      .issues({sort: 'UPDATED_AT'}).
+//      .fetchAll({per_page: 100});
+// 2. fetchAllReactions = await Client.getGraphQLClient()
+//      .repo(repoOwner, repoName)
+//      .reactions({pr_number, per_review: 20})
+//      .fetchOne({per_page: 100});
+
+class GraphQLClientBase {
+  constructor(token, sleepTime = 3, warningThreshold = 15) {
+    this.token = token;
+    this.GH_GQL_BASE = 'https://api.github.com/graphql';
+    this.GH_GQL_OPTIONS = {
+      url: this.GH_GQL_BASE,
+      headers: token
+        ? { Authorization: `bearer ${token}` }
+        : {}
+    };
+    this.client = GraphQL(this.GH_GQL_OPTIONS);
+
+    // sleep time (sec) before timeout retry
+    this.sleepTime = sleepTime;
+
+    // maximum number of warnings before stop fetching
+    this.warningThreshold = warningThreshold;
+  }
+
+  repo(repoOwner, repoName) {
+    this.repoOwner = repoOwner;
+    this.repoName = repoName;
+    if (DEBUG) {
+      console.log('repoOwner set:', repoOwner, 'repoName set:', repoName);
+    }
+    return this;
+  }
+
+  // First type of queries: ISSUES
+  issues(config) {
+    const { sort, direction, earliestDate } = config || {};
+    // There are three types of order
+    // https://developer.github.com/v4/enum/issueorderfield/
+    // COMMENTS, CREATED_AT (default), and UPDATED_AT
+    this.orderBy = {
+      field: sort || 'CREATED_AT',
+      direction: direction || 'ASC'
+    };
+    this._fetch = this._fetchIssues;
+    this.earliestDate = earliestDate;
+    return this;
+  }
+
+  // Second type of queries: PULLREQUESTS
+  pullRequests(config) {
+    const { sort, direction, earliestDate,
+      ignoreAuthor, ignoreContent } = config || {};
+    // order of pull requests and issues follow the same rule
+    // see comment within `issues` method
+    this.orderBy = {
+      field: sort || 'CREATED_AT',
+      direction: direction || 'ASC'
+    };
+    this._fetch = this._fetchPullRequests;
+    this.earliestDate = earliestDate;
+
+    // review will be ignored if it matches any string specified here
+    // it can be a literal or a regular expression
+    // delimiter: space
+    if (ignoreAuthor || ignoreAuthor === '') {
+      this.ignoreAuthor = ignoreAuthor;
+    } else {
+      // default value
+      this.ignoreAuthor = 'gitmate-bot rultor';
+    }
+
+    // review will be ignored if it's written by any author specified here
+    // delimiter: space
+    if (ignoreContent || ignoreContent === '') {
+      this.ignoreContent = ignoreContent;
+    } else {
+      // default value
+      this.ignoreContent = '@gitmate-bot @rultor /^(unack|ack)/g';
+    }
+
+    return this;
+  }
+
+  // Third type of queries: LABELS
+  labels() {
+    this._fetch = this._fetchLabels;
+    return this;
+  }
+
+  // Fourth type of queries: REACTIONS
+  reactions(config) {
+    const { pr_number, reviews_count,
```

Comment:
per consts in `pullRequests(config)`

Reply:
You mean those config (`reviews_count`, `comments_count`, `discussions_per_review`) should be pullRequest config?

That doesn't work in this way. Although the name is `reactions`, it is actually fetching pr reviews (together with reactions), since there is no way to get all reactions directly. To reduce API hits, we don't need to fetch a same number of reviews per PR (say, 20 per PR) when fetching reactions, coz that number has been known when fetching pr reviews.